src/js/media-tooltip.ts

```diff
@@ -0,0 +1,278 @@
+    // tooltip compared to the left edge of the container element, to see if it
+    // bleeds out (and the same for the right edges).
+    // If they do, then we apply the diff as an offset to get it back within bounds
+    // + any extra margin specified to create some buffer space, so it looks better.
+    // e.g. it's 20px out of bounds, we nudge it 20px back in + margin
+    const tooltipStyle = getComputedStyle(this);
+    const containingSelector = this.containWithin ?? defaultContainerSelector;
+    const containingEl = closestComposedNode(this, containingSelector);
+    if (!containingEl) return;
+    const { x: containerX, width: containerWidth } =
+      containingEl.getBoundingClientRect();
+    const { x: tooltipX, width: tooltipWidth } = this.getBoundingClientRect();
+    const tooltipRight = tooltipX + tooltipWidth;
+    const containerRight = containerX + containerWidth;
+    const offsetXVal = tooltipStyle.getPropertyValue(
+      '--media-tooltip-offset-x'
+    );
+    const currOffsetX = offsetXVal
+      ? parseFloat(offsetXVal.replace('px', ''))
+      : 0;
+    const marginVal = tooltipStyle.getPropertyValue(
+      '--media-tooltip-container-margin'
+    );
+    const currMargin = marginVal ? parseFloat(marginVal.replace('px', '')) : 0;
+
+    // We might have already offset the tooltip previously so we remove it's
+    // current offset from our calculations, because we need to know if it goes
+    // outside the boundary if we weren't already offsetting it
+    // We also add on any additional container margin specified. Depending on
+    // if we're adjusting the element leftwards or rightwards, we need either a
+    // positive or negative offset
+    const leftDiff = tooltipX - containerX + currOffsetX - currMargin;
+    const rightDiff = tooltipRight - containerRight + currOffsetX + currMargin;
+
+    // out of left bounds
+    if (leftDiff < 0) {
+      this.style.setProperty('--media-tooltip-offset-x', `${leftDiff}px`);
+      return;
+    }
+
+    // out of right bounds
+    if (rightDiff > 0) {
+      this.style.setProperty('--media-tooltip-offset-x', `${rightDiff}px`);
+      return;
+    }
+
+    // no spilling out
+    this.style.removeProperty('--media-tooltip-offset-x');
+  };
+
+  /**
+   * Get or set tooltip position
+   */
+  get position(): TooltipPosition | undefined {
+    return getStringAttr(this, Attributes.POSITION);
+  }
+
+  set position(value: TooltipPosition | undefined) {
+    setStringAttr(this, Attributes.POSITION, value);
+  }
+
+  /**
+   * Get or set tooltip container selector that will constrain the tooltips horizontal
+   * position. A css selector that matches one of the tooltips parents.
+   */
+  get containWithin(): string | undefined {
+    return getStringAttr(this, Attributes.CONTAIN_WITHIN);
+  }
+
+  set containWithin(value: string | undefined) {
+    setStringAttr(this, Attributes.CONTAIN_WITHIN, value);
+  }
```

Comment:
see other comment, we should try to align on this naming. we've used `bounds` in other elements

Reply:
Just resolving this convo in favour of the other one above